lib/pulse_core/PulseCore.Atomic.fst

```diff
@@ -0,0 +1,313 @@
+
+let bind_atomic
+    (#a:Type u#a)
+    (#b:Type u#b)
+    (#obs1:_)
+    (#obs2:observability { at_most_one_observable obs1 obs2 })
+    (#opens:inames)
+    (#pre1:slprop)
+    (#post1:a -> slprop)
+    (#post2:b -> slprop)
+    (e1:stt_atomic a #obs1 opens pre1 post1)
+    (e2:(x:a -> stt_atomic b #obs2 opens (post1 x) post2))
+= A.bind e1 e2
+
+let lift_unobservable
+    (#a:Type u#a)
+    (#obs:_)
+    (#opens:inames)
+    (#pre:slprop)
+    (#post:a -> slprop)
+    (e:stt_atomic a #obs opens pre post)
+= e
+
+let lift_atomic0
+    (#a:Type u#0)
+    (#obs:_)
+    (#opens:inames)
+    (#pre:slprop)
+    (#post:a -> slprop)
+    (e:stt_atomic a #obs opens pre post)
+: stt a pre post
+= A.lift0 e
+
+let lift_atomic1
+    (#a:Type u#1)
+    (#obs:_)
+    (#opens:inames)
+    (#pre:slprop)
+    (#post:a -> slprop)
+    (e:stt_atomic a #obs opens pre post)
+: stt a pre post
+= A.lift1 e
+
+let lift_atomic2
+    (#a:Type u#2)
+    (#obs:_)
+    (#opens:inames)
+    (#pre:slprop)
+    (#post:a -> slprop)
+    (e:stt_atomic a #obs opens pre post)
+: stt a pre post
+= A.lift2 e
+
+let frame_atomic
+    (#a:Type u#a)
+    (#obs: observability)
+    (#opens:inames)
+    (#pre:slprop)
+    (#post:a -> slprop)
+    (frame:slprop)
+    (e:stt_atomic a #obs opens pre post)
+: stt_atomic a #obs opens (pre ** frame) (fun x -> post x ** frame)
+= A.frame e
+
+let sub_atomic
+    (#a:Type u#a)
+    (#obs:_)
+    (#opens:inames)
+    (#pre1:slprop)
+    (pre2:slprop)
+    (#post1:a -> slprop)
+    (post2:a -> slprop)
+    (pf1 : slprop_equiv pre1 pre2)
+    (pf2 : slprop_post_equiv post1 post2)
+    (e:stt_atomic a #obs opens pre1 post1)
+: stt_atomic a #obs opens pre2 post2
+= A.sub pre2 post2 e
+
+let sub_invs_stt_atomic
+    (#a:Type u#a)
+    (#obs:_)
+    (#opens1 #opens2:inames)
+    (#pre:slprop)
+    (#post:a -> slprop)
+    (e:stt_atomic a #obs opens1 pre post)
+    (_ : squash (inames_subset opens1 opens2))
+: stt_atomic a #obs opens2 pre post
+= assert (Set.equal (Set.union opens1 opens2) opens2);
+  A.weaken opens2 e
+
+let stt_ghost a opens pre post = Ghost.erased (act a opens pre post)
+let return_ghost #a x p = Ghost.hide (return_atomic #a x p)
+let return_ghost_noeq #a x p = Ghost.hide (A.return #_ #p x)
+let bind_ghost
+    (#a:Type u#a)
+    (#b:Type u#b)
+    (#opens:inames)
+    (#pre1:slprop)
+    (#post1:a -> slprop)
+    (#post2:b -> slprop)
+    (e1:stt_ghost a opens pre1 post1)
+    (e2:(x:a -> stt_ghost b opens (post1 x) post2))
+: stt_ghost b opens pre1 post2
+= let e1 = Ghost.reveal e1 in
+  let e2 = FStar.Ghost.Pull.pull (fun (x:a) -> Ghost.reveal (e2 x)) in
+  Ghost.hide (A.bind e1 e2)
+
+let lift_ghost
+    (#a:Type u#a)
+    (#opens:inames)
+    (#pre:slprop)
+    (#post:a -> slprop)
+    (e:stt_ghost a opens pre post)
+    (reveal_a:non_informative_witness a)
+: stt_atomic a #Unobservable opens pre post
+= admit() //This is the main axiom about ghost computations; in Steel, this axiom is implemented within the effect system

```

Comment:
In Steel, we mark Steel.Effect.STGhost as erasable, which instructs the F* typechecker and effect system to allow lifts to non-erasable effects if the result is non-informative. PulseCore is not using the effect system, so the axiom is manifested explicitly here.

Reply:
Ah.. I now see your comment about forbidding concrete writes.